File: java/proverb/src/main/java/Proverb.java
class Proverb {

    private final String[] words;

    Proverb(String[] words) {
        this.words = words;
    }

    String recite() {
        if (words.length == 0) return "";

        StringBuilder proverb = new StringBuilder();
        for (int i = 0; i < words.length - 1; i++) {
            proverb
                .append("For want of a ")
                .append(words[i])
                .append(" the ")
                .append(words[i + 1])
                .append(" was lost.\n");
        }
        return proverb + "And all for the want of a " + words[0] + ".";
    }

}

File: go/proverb/proverb.go
// Package proverb generates a wide range of pithy proverbs (currently one).
package proverb

import "fmt"

// Proverb returns the "For Want of a Nail" proverb.
func Proverb(rhyme []string) (result []string) {
	numWords := len(rhyme)
	result = make([]string, numWords)

	for i := 0; i < numWords-1; i++ {
		result[i] = fmt.Sprintf("For want of a %s the %s was lost.", rhyme[i], rhyme[i+1])
	}
	if numWords > 0 {
		result[numWords-1] = fmt.Sprintf("And all for the want of a %s.", rhyme[0])
	}
	return
}
